usage/rest-client/src/main/java/brooklyn/rest/client/BrooklynApi.java

```diff
@@ -116,22 +120,37 @@ public BrooklynApi(URL endpoint, ClientExecutor clientExecutor) {
         final T result0 = ProxyFactory.create(clazz, target, clientExecutor);
         return (T) Proxy.newProxyInstance(clazz.getClassLoader(), new Class<?>[] { clazz }, new InvocationHandler() {
             @Override
-            public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
-                Object result1;
+            public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {                 
                 try {
-                    result1 = method.invoke(result0, args);
+                    Object result1 = method.invoke(result0, args);
+                    Class<?> type = String.class;
+                    if (result1 instanceof Response) {
+                        Response resp = (Response)result1;
+                        if(HttpTool.isStatusCodeHealthy(resp.getStatus()) && method.isAnnotationPresent(ApiOperation.class)) {
+                           type = getClassFromMethodAnnotationOrDefault(resp, method, type);
+                        }
+                        // wrap the original response so it self-closes
+                        result1 = BuiltResponsePreservingError.copyResponseAndClose(resp, type);
+                    }
+                    return result1;
                 } catch (Throwable e) {
                     if (e instanceof InvocationTargetException) {
                         // throw the original exception
                         e = ((InvocationTargetException)e).getTargetException();
                     }
                     throw Exceptions.propagate(e);
+                }  
+            }
+            
+            private Class<?> getClassFromMethodAnnotationOrDefault(Response resp, Method method, Class<?> def){
+                Class<?> type;
+                try{
+                    String responseClass = method.getAnnotation(ApiOperation.class).responseClass();
+                    type = Class.forName(responseClass);
+                } catch (Exception e) {
+                    type = def;
```

Comment:
Worth a log statement, rather than swallowing the exception, to explain (in the debug log) why it defaulted to a String?

I also like to do `Exceptions.propagateIfFatal()`. This mainly is for handing InterruptedExceptions and RuntimeInterruptedExceptions right, which should never happen in the calls within the try block (so not important here).